### DISCUSSION TITLE
Hot Corner option to display windows of application

### DISCUSSION THREAD

**Discussion by GalacticSun** [linuxmint/discussions#410] [Ideas]
Hi there,

Would it be possible to have an option for Hot Corners that displays the windows of only the current/active application?  For example, if you're using your web browser and activate this Hot Corner, it would display just the browser's windows.  It would be really handy for those of us who have too many browser windows open at a time.

I bring this up because I used to use Mac OS and loved their "Exposé" / "Mission Control" feature, which is just like Linux Mint's Hot Corners->Show All Windows feature.  But Mac OS also has a slightly modified version of this feature that allows the user to show the windows of only the current application you're using. They have it available as a Hot Corner option called "Application Windows".

Thanks very much and keep up the great work.

Related:
https://www.digitaltrends.com/computing/how-to-use-mac-hot-corners/
https://osxdaily.com/2021/06/24/use-hot-corners-mac/
And:
https://forums.linuxmint.com/viewtopic.php?t=379430
https://github.com/linuxmint/cinnamon/issues/2117

**Comment by Jopp-gh** [comment]
Mint ships with 2 applets, 
1. to see the windows of apps, 
2. to see your workspaces. 

Just activate both in **system settings**.
Back to your panel, right click on them and make them open on hover.

**Comment by GalacticSun** [comment]
Thanks, @Jopp-gh, but, unfortunately, those don't have the functionality I'm looking for, as far as I can tell.  I've tried the following applets in the panel:

- Expo (displays workspaces, not windows of one/the active application)
- Scale (this does show the tiling effect, but it displays the windows of all applications, not just windows of one/the active application.  This is the closest I could find, but is the same as the current hot corner functionality which shows all windows, not the windows of the active application only)
- Grouped Window List (displays all windows in a list coming up from the panel, but not as tiles like the hot corner option, and doesn't display just the windows of one/the active application - and for my browser there are far too many windows - 88 at the moment, and the lists goes off the top of my screen - I'm looking for the hot corner tiled effect)
- Window List (displays all windows in the panel, not as windows of one/the active application - and there are far too many windows to display, so they appear as small vertical lines, and they are persistent)
- Corner Bar (Shows all  windows, but doesn't show the windows of one/the active application)

If I am missing the ones you are referring to, can you be more specific, please?  Otherwise, there are no applets that provide the functionality I'm describing.

**Reply by Jopp-gh** [reply]
I'm happy with the default core applets and didn't know about user made applets for app window management.
Same, I was a Mac user for decades. Mint is a bit different, so my suggestion is to always use tabs if possible, combined with workspaces.
To isolate the current app you'd need to write a bash script, to minimize all other app windows.

**Comment by anaximeno** [comment]
For now you can achieve that by hitting `Ctrl+\` on your keyboard, and it will only show you the windows from your current application on the windows switcher.

**Reply by Jopp-gh** [reply]
Just for curiosity, I tried this shortcut but it doesn't seem to work for me, maybe because of my keyboard layout. (Mint Virginia 21.3) Also, I didn't find this shortcut under `system settings>keyboard>shortcuts`
The setting I found was a circular scrolling through windows of the same app. That's nice.

I think `scale@cinnamon` and `expo@cinnamon` are great, bc both can be activated via hover. 

A window overview of the current app isn't a bad idea, but I hope that Mint will integrate tabs in the near future. Multiple windows for each app is OK but just a mess. Tabs like in a browser are easier to manage.

**Comment by GalacticSun** [comment]
Thanks, although, I'm actually meaning this as a Feature Request for the developers.  I read that this is the correct place for that, and not to post this under "Issues" here on Github.  I'm hoping that this can be seen as a Feature Request.  Or, if a LM / Cinnamon developer can point me to where the code is for this, I can try to modify the code myself.

**Reply by anaximeno** [reply]
Yeah, I've suggested `Ctrl+\` as an alternative while this isn't implemented. I suspect this is implemented at https://github.com/linuxmint/cinnamon, but I must confirm that. You can try to look into it if you are willing to. You can also try to see there how the `Ctrl+\` is implemented to see how to find and sort windows only from the same app, and implement that with something similar to the expo view and add that option to the choice of the available actions on the hot corner program.

**Reply by GalacticSun** [reply]
I've looked into it a bit, but I couldn't find out where exactly the function call is or what is even responsible for the tiling animation, much less where the windows are retrieved from, etc.  So I'd need a bit of specific guidance as to which part of which file to change.  Btw, `Ctrl+\` doesn't work for me either, likely a keyboard layout thing, I suppose (I'm using English US).

**Comment by GalacticSun** [comment]
Can a Cinnamon/LM developer point me in the right direction here, please?  Where is the code that I might be able to edit to make this change?

**Comment by GalacticSun** [comment]
So is there no hope of a response, then?  I'm happy to try and make a change in the code myself if someone can tell me where to make it.  Anyone?

**Reply by cplittlfield** [reply]
You may need to write an applet to do what you want.  Why not start with the code for Hot Corners, and work from there?

Myself, I happy that I can achieve the sorts of things you describe by using Alt+Tab to cycle through all open windows on the workspace, and Alt+Esc to cycle through the windows (not tabs) of the active application.  I also have the Grouped Windows applet set to display icons on the panel for all open apps (instead of just the active workspace), and to show snapshots of all the open windows for any given application when I hover the pointer over its panel icon.  But I sympathize with your desire to make it all work the way you are accustomed to.

**Reply by GalacticSun** [reply]
> Why not start with the code for Hot Corners, and work from there?

Yes, that's what I'm asking: where is this code in the Cinnamon codebase?  I can't find it.  And no developers appear to want to actually help me, I guess they don't read these posts.  I'd like to create another option within Hot Corners directly, though, not create an applet.

**Reply by cplittlfield** [reply]
Take a look at the following:
/usr/share/cinnamon/cinnamon-settings/modules/__pycache__/cs_hotcorner.cpython-310.pyc
/usr/share/cinnamon/cinnamon-settings/modules/cs_hotcorner.py

**Reply by GalacticSun** [reply]
Thanks, yeah, I did actually find that `cs_hotcorner.py` file by searching the repo, but I can't find the code for the `scale` function, which appears to be what I'm looking for (I _think_).  `cs_hotcorner.py` appears to handle the setup and management of the options when setting up Hotcorners in the Cinnamon settings, etc.  I'm trying to find the code that does the actual animation and control of the `scale` effect: when all the windows become small and tiled, after you trigger that hotcorner effect.  That's the code I can't find - that's exactly where I'm struggling with this.  I appreciate your help.

**Reply by cplittlfield** [reply]
Sorry I couldn't help more, but I am officially way out of my depth!

Good luck! :smile: :+1:
